
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Benchmarking Hub for Genetics & Brain Foundation Models">
      
      
      
        <link rel="canonical" href="https://allison-eunse.github.io/ai4h-inspired-fm-benchmark-hub/integration/modality_features/genomics/">
      
      
        <link rel="prev" href="../smri/">
      
      
        <link rel="next" href="../../../models/">
      
      
        
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.7.0">
    
    
      
        <title>Genomics - AI4H-Inspired FM Benchmarks</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.618322db.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.ab4e12ef.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Inter:300,300i,400,400i,700,700i%7CJetBrains+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Inter";--md-code-font:"JetBrains Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="deep-purple">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#genomics-data-specifications" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="AI4H-Inspired FM Benchmarks" class="md-header__button md-logo" aria-label="AI4H-Inspired FM Benchmarks" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M21.33 12.91c.09 1.55-.62 3.04-1.89 3.95l.77 1.49c.23.45.26.98.06 1.45-.19.47-.58.84-1.06 1l-.79.25a1.687 1.687 0 0 1-1.86-.55L14.44 18c-.89-.15-1.73-.53-2.44-1.1-.5.15-1 .23-1.5.23-.88 0-1.76-.27-2.5-.79-.53.16-1.07.23-1.62.22-.79.01-1.57-.15-2.3-.45a4.1 4.1 0 0 1-2.43-3.61c-.08-.72.04-1.45.35-2.11-.29-.75-.32-1.57-.07-2.33C2.3 7.11 3 6.32 3.87 5.82c.58-1.69 2.21-2.82 4-2.7 1.6-1.5 4.05-1.66 5.83-.37.42-.11.86-.17 1.3-.17 1.36-.03 2.65.57 3.5 1.64 2.04.53 3.5 2.35 3.58 4.47.05 1.11-.25 2.2-.86 3.13.07.36.11.72.11 1.09m-5-1.41c.57.07 1.02.5 1.02 1.07a1 1 0 0 1-1 1h-.63c-.32.9-.88 1.69-1.62 2.29.25.09.51.14.77.21 5.13-.07 4.53-3.2 4.53-3.25a2.59 2.59 0 0 0-2.69-2.49 1 1 0 0 1-1-1 1 1 0 0 1 1-1c1.23.03 2.41.49 3.33 1.3.05-.29.08-.59.08-.89-.06-1.24-.62-2.32-2.87-2.53-1.25-2.96-4.4-1.32-4.4-.4-.03.23.21.72.25.75a1 1 0 0 1 1 1c0 .55-.45 1-1 1-.53-.02-1.03-.22-1.43-.56-.48.31-1.03.5-1.6.56-.57.05-1.04-.35-1.07-.9a.97.97 0 0 1 .88-1.1c.16-.02.94-.14.94-.77 0-.66.25-1.29.68-1.79-.92-.25-1.91.08-2.91 1.29C6.75 5 6 5.25 5.45 7.2 4.5 7.67 4 8 3.78 9c1.08-.22 2.19-.13 3.22.25.5.19.78.75.59 1.29-.19.52-.77.78-1.29.59-.73-.32-1.55-.34-2.3-.06-.32.27-.32.83-.32 1.27 0 .74.37 1.43 1 1.83.53.27 1.12.41 1.71.4q-.225-.39-.39-.81a1.038 1.038 0 0 1 1.96-.68c.4 1.14 1.42 1.92 2.62 2.05 1.37-.07 2.59-.88 3.19-2.13.23-1.38 1.34-1.5 2.56-1.5m2 7.47-.62-1.3-.71.16 1 1.25zm-4.65-8.61a1 1 0 0 0-.91-1.03c-.71-.04-1.4.2-1.93.67-.57.58-.87 1.38-.84 2.19a1 1 0 0 0 1 1c.57 0 1-.45 1-1 0-.27.07-.54.23-.76.12-.1.27-.15.43-.15.55.03 1.02-.38 1.02-.92"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            AI4H-Inspired FM Benchmarks
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Genomics
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="deep-purple"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="deep-purple"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12s-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/allison-eunse/ai4h-inspired-fm-benchmark-hub" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    ai4h-inspired-fm-benchmark-hub
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../../start_here/" class="md-tabs__link">
        
  
  
    
  
  Start Here

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../../how_it_works/" class="md-tabs__link">
        
  
  
    
  
  How It Works

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../../leaderboards/" class="md-tabs__link">
        
  
  
    
  
  Leaderboards

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../contributing/submission_guide/" class="md-tabs__link">
          
  
  
  Submit Results

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../analysis_recipes/cca_permutation/" class="md-tabs__link">
          
  
  
  Protocols (Recipes)

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../../data_sources/" class="md-tabs__link">
          
  
  
  Data Specs

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../models/" class="md-tabs__link">
          
  
  
  Models

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../design/ai4h_alignment/" class="md-tabs__link">
          
  
  
  Design

        </a>
      </li>
    
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../../credits/" class="md-tabs__link">
        
  
  
    
  
  Credits

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="AI4H-Inspired FM Benchmarks" class="md-nav__button md-logo" aria-label="AI4H-Inspired FM Benchmarks" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M21.33 12.91c.09 1.55-.62 3.04-1.89 3.95l.77 1.49c.23.45.26.98.06 1.45-.19.47-.58.84-1.06 1l-.79.25a1.687 1.687 0 0 1-1.86-.55L14.44 18c-.89-.15-1.73-.53-2.44-1.1-.5.15-1 .23-1.5.23-.88 0-1.76-.27-2.5-.79-.53.16-1.07.23-1.62.22-.79.01-1.57-.15-2.3-.45a4.1 4.1 0 0 1-2.43-3.61c-.08-.72.04-1.45.35-2.11-.29-.75-.32-1.57-.07-2.33C2.3 7.11 3 6.32 3.87 5.82c.58-1.69 2.21-2.82 4-2.7 1.6-1.5 4.05-1.66 5.83-.37.42-.11.86-.17 1.3-.17 1.36-.03 2.65.57 3.5 1.64 2.04.53 3.5 2.35 3.58 4.47.05 1.11-.25 2.2-.86 3.13.07.36.11.72.11 1.09m-5-1.41c.57.07 1.02.5 1.02 1.07a1 1 0 0 1-1 1h-.63c-.32.9-.88 1.69-1.62 2.29.25.09.51.14.77.21 5.13-.07 4.53-3.2 4.53-3.25a2.59 2.59 0 0 0-2.69-2.49 1 1 0 0 1-1-1 1 1 0 0 1 1-1c1.23.03 2.41.49 3.33 1.3.05-.29.08-.59.08-.89-.06-1.24-.62-2.32-2.87-2.53-1.25-2.96-4.4-1.32-4.4-.4-.03.23.21.72.25.75a1 1 0 0 1 1 1c0 .55-.45 1-1 1-.53-.02-1.03-.22-1.43-.56-.48.31-1.03.5-1.6.56-.57.05-1.04-.35-1.07-.9a.97.97 0 0 1 .88-1.1c.16-.02.94-.14.94-.77 0-.66.25-1.29.68-1.79-.92-.25-1.91.08-2.91 1.29C6.75 5 6 5.25 5.45 7.2 4.5 7.67 4 8 3.78 9c1.08-.22 2.19-.13 3.22.25.5.19.78.75.59 1.29-.19.52-.77.78-1.29.59-.73-.32-1.55-.34-2.3-.06-.32.27-.32.83-.32 1.27 0 .74.37 1.43 1 1.83.53.27 1.12.41 1.71.4q-.225-.39-.39-.81a1.038 1.038 0 0 1 1.96-.68c.4 1.14 1.42 1.92 2.62 2.05 1.37-.07 2.59-.88 3.19-2.13.23-1.38 1.34-1.5 2.56-1.5m2 7.47-.62-1.3-.71.16 1 1.25zm-4.65-8.61a1 1 0 0 0-.91-1.03c-.71-.04-1.4.2-1.93.67-.57.58-.87 1.38-.84 2.19a1 1 0 0 0 1 1c.57 0 1-.45 1-1 0-.27.07-.54.23-.76.12-.1.27-.15.43-.15.55.03 1.02-.38 1.02-.92"/></svg>

    </a>
    AI4H-Inspired FM Benchmarks
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/allison-eunse/ai4h-inspired-fm-benchmark-hub" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M173.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M252.8 8C114.1 8 8 113.3 8 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C436.2 457.8 504 362.9 504 252 504 113.3 391.5 8 252.8 8M105.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    ai4h-inspired-fm-benchmark-hub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Home
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../start_here/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Start Here
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../how_it_works/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    How It Works
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../leaderboards/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Leaderboards
  

    
  </span>
  
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Submit Results
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            
  
    Submit Results
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../contributing/submission_guide/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Submission Guide
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Protocols (Recipes)
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            
  
    Protocols (Recipes)
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../analysis_recipes/cca_permutation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    CCA & Permutation
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../analysis_recipes/prediction_baselines/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Prediction Baselines
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../analysis_recipes/partial_correlations/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Partial Correlations
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_7" checked>
        
          
          <label class="md-nav__link" for="__nav_7" id="__nav_7_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    
  
    Data Specs
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_7_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_7">
            <span class="md-nav__icon md-icon"></span>
            
  
    Data Specs
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../data_sources/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Free Data Sources
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../fmri/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    fMRI
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../smri/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    sMRI
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    
  
    Genomics
  

    
  </span>
  
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    
  
    Genomics
  

    
  </span>
  
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      
        Overview
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-modalities" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data Modalities
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Data Modalities">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-single-cell-rna-seq-scrna-seq" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Single-Cell RNA-seq (scRNA-seq)
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Single-Cell RNA-seq (scRNA-seq)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#required-metadata-adataobs" class="md-nav__link">
    <span class="md-ellipsis">
      
        Required Metadata (adata.obs)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene-metadata-adatavar" class="md-nav__link">
    <span class="md-ellipsis">
      
        Gene Metadata (adata.var)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-bulk-rna-seq" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. Bulk RNA-seq
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-dna-sequences" class="md-nav__link">
    <span class="md-ellipsis">
      
        3. DNA Sequences
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-variant-data-vcf" class="md-nav__link">
    <span class="md-ellipsis">
      
        4. Variant Data (VCF)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#preprocessing-workflows" class="md-nav__link">
    <span class="md-ellipsis">
      
        Preprocessing Workflows
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Preprocessing Workflows">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#scrna-seq-standard-preprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      
        scRNA-seq Standard Preprocessing
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#bulk-rna-seq-preprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      
        Bulk RNA-seq Preprocessing
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#dna-sequence-preprocessing" class="md-nav__link">
    <span class="md-ellipsis">
      
        DNA Sequence Preprocessing
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-augmentation-for-scrna-seq" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data Augmentation for scRNA-seq
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Data Augmentation for scRNA-seq">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-poisson-sampling-simulates-sequencing-depth-variation" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Poisson Sampling (Simulates sequencing depth variation)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-cell-mixtures-simulates-doublets" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. Cell Mixtures (Simulates doublets)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#benchmark-tasks" class="md-nav__link">
    <span class="md-ellipsis">
      
        Benchmark Tasks
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Benchmark Tasks">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-cell-type-annotation-scrna-seq" class="md-nav__link">
    <span class="md-ellipsis">
      
        1. Cell Type Annotation (scRNA-seq)
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-gene-expression-prediction" class="md-nav__link">
    <span class="md-ellipsis">
      
        2. Gene Expression Prediction
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-variant-effect-prediction" class="md-nav__link">
    <span class="md-ellipsis">
      
        3. Variant Effect Prediction
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#4-sequence-classification" class="md-nav__link">
    <span class="md-ellipsis">
      
        4. Sequence Classification
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#quality-control-metrics" class="md-nav__link">
    <span class="md-ellipsis">
      
        Quality Control Metrics
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Quality Control Metrics">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#scrna-seq-qc" class="md-nav__link">
    <span class="md-ellipsis">
      
        scRNA-seq QC
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#variant-data-qc" class="md-nav__link">
    <span class="md-ellipsis">
      
        Variant Data QC
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-formats" class="md-nav__link">
    <span class="md-ellipsis">
      
        Data Formats
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Data Formats">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#anndata-structure" class="md-nav__link">
    <span class="md-ellipsis">
      
        AnnData Structure
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-dataset-metadata" class="md-nav__link">
    <span class="md-ellipsis">
      
        Example Dataset Metadata
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#foundation-model-input-formats" class="md-nav__link">
    <span class="md-ellipsis">
      
        Foundation Model Input Formats
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Foundation Model Input Formats">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#for-geneformer-and-similar-models" class="md-nav__link">
    <span class="md-ellipsis">
      
        For Geneformer and Similar Models
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#for-dna-sequence-models-dnabert-hyenadna-etc" class="md-nav__link">
    <span class="md-ellipsis">
      
        For DNA Sequence Models (DNABERT, HyenaDNA, etc.)
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#itu-ai4h-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      
        ITU AI4H Alignment
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tools-libraries" class="md-nav__link">
    <span class="md-ellipsis">
      
        Tools &amp; Libraries
      
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Tools &amp; Libraries">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#scrna-seq" class="md-nav__link">
    <span class="md-ellipsis">
      
        scRNA-seq
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genomics" class="md-nav__link">
    <span class="md-ellipsis">
      
        Genomics
      
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#foundation-model-libraries" class="md-nav__link">
    <span class="md-ellipsis">
      
        Foundation Model Libraries
      
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#references" class="md-nav__link">
    <span class="md-ellipsis">
      
        References
      
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#related-documentation" class="md-nav__link">
    <span class="md-ellipsis">
      
        Related Documentation
      
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_8" >
        
          
          <label class="md-nav__link" for="__nav_8" id="__nav_8_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Models
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_8_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_8">
            <span class="md-nav__icon md-icon"></span>
            
  
    Models
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../models/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Overview
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_9" >
        
          
          <label class="md-nav__link" for="__nav_9" id="__nav_9_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    
  
    Design
  

    
  </span>
  
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_9_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_9">
            <span class="md-nav__icon md-icon"></span>
            
  
    Design
  

          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../design/ai4h_alignment/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    AI4H Alignment
  

    
  </span>
  
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../credits/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    
  
    Credits
  

    
  </span>
  
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="genomics-data-specifications">Genomics Data Specifications</h1>
<h2 id="overview">Overview</h2>
<p>This page defines standardized data formats and preprocessing requirements for genomics data in the benchmark hub, covering single-cell RNA-seq (scRNA-seq), bulk RNA-seq, DNA sequences, and variant data.</p>
<h2 id="data-modalities">Data Modalities</h2>
<h3 id="1-single-cell-rna-seq-scrna-seq">1. Single-Cell RNA-seq (scRNA-seq)</h3>
<p><strong>Format</strong>: AnnData (<code>.h5ad</code>), Loom (<code>.loom</code>), or CSV/TSV matrices<br />
<strong>Shape</strong>: <code>(n_cells, n_genes)</code><br />
<strong>Data Type</strong>: Raw counts (integer) or normalized (float)</p>
<div class="highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="c1"># Load single-cell data</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;pbmc_68k.h5ad&#39;</span><span class="p">)</span>

<span class="c1"># Inspect</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Shape: </span><span class="si">{</span><span class="n">adata</span><span class="o">.</span><span class="n">shape</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>  <span class="c1"># (n_cells, n_genes)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Genes: </span><span class="si">{</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Cells: </span><span class="si">{</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">[:</span><span class="mi">5</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="c1"># Access count matrix</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>  <span class="c1"># Sparse or dense matrix (n_cells, n_genes)</span>
</code></pre></div>
<h4 id="required-metadata-adataobs">Required Metadata (adata.obs)</h4>
<div class="highlight"><pre><span></span><code><span class="c1"># Cell-level metadata</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;cell_type&#39;</span><span class="p">]</span>  <span class="c1"># Cell type annotations</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;donor_id&#39;</span><span class="p">]</span>  <span class="c1"># Subject/donor identifier</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;batch&#39;</span><span class="p">]</span>  <span class="c1"># Batch/experiment ID</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_genes&#39;</span><span class="p">]</span>  <span class="c1"># Number of detected genes</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_counts&#39;</span><span class="p">]</span>  <span class="c1"># Total UMI counts</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;percent_mito&#39;</span><span class="p">]</span>  <span class="c1"># % mitochondrial gene expression</span>
</code></pre></div>
<h4 id="gene-metadata-adatavar">Gene Metadata (adata.var)</h4>
<div class="highlight"><pre><span></span><code><span class="c1"># Gene-level metadata</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;gene_ids&#39;</span><span class="p">]</span>  <span class="c1"># Ensembl IDs</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;gene_symbols&#39;</span><span class="p">]</span>  <span class="c1"># HGNC symbols</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;highly_variable&#39;</span><span class="p">]</span>  <span class="c1"># Boolean for HVG selection</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;mean_counts&#39;</span><span class="p">]</span>  <span class="c1"># Mean expression</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;dispersions&#39;</span><span class="p">]</span>  <span class="c1"># Dispersion (variability)</span>
</code></pre></div>
<h3 id="2-bulk-rna-seq">2. Bulk RNA-seq</h3>
<p><strong>Format</strong>: NumPy array, CSV, or TSV<br />
<strong>Shape</strong>: <code>(n_samples, n_genes)</code><br />
<strong>Data Type</strong>: Raw counts, TPM, or FPKM</p>
<div class="highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="c1"># Load bulk RNA-seq data</span>
<span class="n">gene_expression</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;bulk_rnaseq_tpm.csv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="c1"># Rows: genes, Columns: samples</span>

<span class="c1"># Transpose to (samples, genes) for ML</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">gene_expression</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">values</span>  <span class="c1"># (n_samples, n_genes)</span>
</code></pre></div>
<h3 id="3-dna-sequences">3. DNA Sequences</h3>
<p><strong>Format</strong>: FASTA (<code>.fasta</code>, <code>.fa</code>), plain text, or tokenized arrays<br />
<strong>Alphabet</strong>: <code>{A, C, G, T, N}</code> (N = ambiguous)</p>
<div class="highlight"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">Bio</span><span class="w"> </span><span class="kn">import</span> <span class="n">SeqIO</span>

<span class="c1"># Load FASTA file</span>
<span class="n">sequences</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;sequences.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
    <span class="n">sequences</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>

<span class="c1"># Tokenize sequences (for transformer models)</span>
<span class="n">vocab</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;A&#39;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span> <span class="s1">&#39;&lt;PAD&gt;&#39;</span><span class="p">:</span> <span class="mi">5</span><span class="p">}</span>

<span class="k">def</span><span class="w"> </span><span class="nf">tokenize_sequence</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">vocab</span><span class="p">,</span> <span class="n">max_len</span><span class="o">=</span><span class="mi">512</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Convert DNA sequence to token IDs.&quot;&quot;&quot;</span>
    <span class="n">tokens</span> <span class="o">=</span> <span class="p">[</span><span class="n">vocab</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">base</span><span class="p">,</span> <span class="n">vocab</span><span class="p">[</span><span class="s1">&#39;N&#39;</span><span class="p">])</span> <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span>

    <span class="c1"># Pad or truncate</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tokens</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">max_len</span><span class="p">:</span>
        <span class="n">tokens</span> <span class="o">+=</span> <span class="p">[</span><span class="n">vocab</span><span class="p">[</span><span class="s1">&#39;&lt;PAD&gt;&#39;</span><span class="p">]]</span> <span class="o">*</span> <span class="p">(</span><span class="n">max_len</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">tokens</span><span class="p">))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">tokens</span> <span class="o">=</span> <span class="n">tokens</span><span class="p">[:</span><span class="n">max_len</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">tokens</span><span class="p">)</span>

<span class="n">tokenized</span> <span class="o">=</span> <span class="p">[</span><span class="n">tokenize_sequence</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">vocab</span><span class="p">)</span> <span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">]</span>
<span class="n">X_dna</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">tokenized</span><span class="p">)</span>  <span class="c1"># (n_sequences, max_len)</span>
</code></pre></div>
<h3 id="4-variant-data-vcf">4. Variant Data (VCF)</h3>
<p><strong>Format</strong>: VCF (Variant Call Format)<br />
<strong>Use</strong>: SNP arrays, whole-genome sequencing (WGS), whole-exome sequencing (WES)</p>
<div class="highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">allel</span>

<span class="c1"># Load VCF file</span>
<span class="n">vcf_path</span> <span class="o">=</span> <span class="s1">&#39;cohort_variants.vcf.gz&#39;</span>
<span class="n">callset</span> <span class="o">=</span> <span class="n">allel</span><span class="o">.</span><span class="n">read_vcf</span><span class="p">(</span><span class="n">vcf_path</span><span class="p">)</span>

<span class="c1"># Extract genotypes</span>
<span class="n">genotypes</span> <span class="o">=</span> <span class="n">callset</span><span class="p">[</span><span class="s1">&#39;calldata/GT&#39;</span><span class="p">]</span>  <span class="c1"># (n_variants, n_samples, ploidy=2)</span>

<span class="c1"># Convert to dosage (0, 1, 2 for diploid)</span>
<span class="n">dosage</span> <span class="o">=</span> <span class="n">genotypes</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>  <span class="c1"># (n_variants, n_samples)</span>

<span class="c1"># Transpose for ML: (samples, variants)</span>
<span class="n">X_geno</span> <span class="o">=</span> <span class="n">dosage</span><span class="o">.</span><span class="n">T</span>
</code></pre></div>
<h2 id="preprocessing-workflows">Preprocessing Workflows</h2>
<h3 id="scrna-seq-standard-preprocessing">scRNA-seq Standard Preprocessing</h3>
<div class="highlight"><pre><span></span><code><span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>

<span class="c1"># Load raw counts</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;raw_counts.h5ad&#39;</span><span class="p">)</span>

<span class="c1"># 1. Quality control</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>  <span class="c1"># Remove low-quality cells</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>  <span class="c1"># Remove rare genes</span>

<span class="c1"># Calculate QC metrics</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;mt&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;MT-&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> 
    <span class="n">qc_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;mt&#39;</span><span class="p">],</span> 
    <span class="n">percent_top</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> 
    <span class="n">log1p</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> 
    <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="c1"># Filter cells by QC</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">n_genes_by_counts</span> <span class="o">&lt;</span> <span class="mi">2500</span><span class="p">,</span> <span class="p">:]</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">pct_counts_mt</span> <span class="o">&lt;</span> <span class="mi">5</span><span class="p">,</span> <span class="p">:]</span>

<span class="c1"># 2. Normalization</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>  <span class="c1"># CPM-like</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>  <span class="c1"># Log-transform</span>

<span class="c1"># 3. Feature selection</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">2000</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">highly_variable</span><span class="p">]</span>  <span class="c1"># Keep only HVGs</span>

<span class="c1"># 4. Scaling (for PCA/visualization, not always for ML)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>

<span class="c1"># Extract preprocessed data</span>
<span class="n">X_processed</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span>  <span class="c1"># (n_cells, 2000)</span>
</code></pre></div>
<h3 id="bulk-rna-seq-preprocessing">Bulk RNA-seq Preprocessing</h3>
<div class="highlight"><pre><span></span><code><span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.preprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">StandardScaler</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="c1"># Load TPM or FPKM data</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">gene_expression</span><span class="o">.</span><span class="n">values</span>  <span class="c1"># (n_samples, n_genes)</span>

<span class="c1"># 1. Log-transform (if not already)</span>
<span class="n">X_log</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>

<span class="c1"># 2. Filter low-expression genes</span>
<span class="n">gene_means</span> <span class="o">=</span> <span class="n">X_log</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">high_expr_genes</span> <span class="o">=</span> <span class="n">gene_means</span> <span class="o">&gt;</span> <span class="mf">1.0</span>  <span class="c1"># Threshold</span>
<span class="n">X_filtered</span> <span class="o">=</span> <span class="n">X_log</span><span class="p">[:,</span> <span class="n">high_expr_genes</span><span class="p">]</span>

<span class="c1"># 3. Z-score normalization (optional)</span>
<span class="n">scaler</span> <span class="o">=</span> <span class="n">StandardScaler</span><span class="p">()</span>
<span class="n">X_scaled</span> <span class="o">=</span> <span class="n">scaler</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X_filtered</span><span class="p">)</span>
</code></pre></div>
<h3 id="dna-sequence-preprocessing">DNA Sequence Preprocessing</h3>
<div class="highlight"><pre><span></span><code><span class="c1"># K-mer encoding (alternative to tokenization)</span>
<span class="k">def</span><span class="w"> </span><span class="nf">generate_kmers</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">3</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Generate k-mer features from DNA sequence.&quot;&quot;&quot;</span>
    <span class="n">kmers</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">-</span> <span class="n">k</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="n">kmers</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequence</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="n">k</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">kmers</span>

<span class="c1"># Build k-mer vocabulary</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">collections</span><span class="w"> </span><span class="kn">import</span> <span class="n">Counter</span>

<span class="n">all_kmers</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">sequences</span><span class="p">:</span>
    <span class="n">all_kmers</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">generate_kmers</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">3</span><span class="p">))</span>

<span class="n">kmer_vocab</span> <span class="o">=</span> <span class="p">{</span><span class="n">kmer</span><span class="p">:</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">kmer</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">all_kmers</span><span class="p">))}</span>

<span class="c1"># Encode sequence as k-mer counts</span>
<span class="k">def</span><span class="w"> </span><span class="nf">encode_kmers</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">kmer_vocab</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">3</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Encode sequence as k-mer count vector.&quot;&quot;&quot;</span>
    <span class="n">kmers</span> <span class="o">=</span> <span class="n">generate_kmers</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">k</span><span class="p">)</span>
    <span class="n">counts</span> <span class="o">=</span> <span class="n">Counter</span><span class="p">(</span><span class="n">kmers</span><span class="p">)</span>

    <span class="n">vec</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">kmer_vocab</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">kmer</span><span class="p">,</span> <span class="n">count</span> <span class="ow">in</span> <span class="n">counts</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">kmer</span> <span class="ow">in</span> <span class="n">kmer_vocab</span><span class="p">:</span>
            <span class="n">vec</span><span class="p">[</span><span class="n">kmer_vocab</span><span class="p">[</span><span class="n">kmer</span><span class="p">]]</span> <span class="o">=</span> <span class="n">count</span>

    <span class="k">return</span> <span class="n">vec</span>
</code></pre></div>
<h2 id="data-augmentation-for-scrna-seq">Data Augmentation for scRNA-seq</h2>
<h3 id="1-poisson-sampling-simulates-sequencing-depth-variation">1. Poisson Sampling (Simulates sequencing depth variation)</h3>
<div class="highlight"><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">poisson_augment</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">factor</span><span class="o">=</span><span class="mf">0.8</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Simulate different sequencing depths via Poisson sampling.</span>

<span class="sd">    Args:</span>
<span class="sd">        counts: (n_cells, n_genes) - raw count matrix</span>
<span class="sd">        factor: Downsampling factor (0 &lt; factor &lt; 1)</span>

<span class="sd">    Returns:</span>
<span class="sd">        augmented: Downsampled count matrix</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">total_counts</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">keepdims</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">target_counts</span> <span class="o">=</span> <span class="n">total_counts</span> <span class="o">*</span> <span class="n">factor</span>

    <span class="c1"># Multinomial sampling</span>
    <span class="n">probs</span> <span class="o">=</span> <span class="n">counts</span> <span class="o">/</span> <span class="p">(</span><span class="n">total_counts</span> <span class="o">+</span> <span class="mf">1e-8</span><span class="p">)</span>
    <span class="n">augmented</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span>
        <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">multinomial</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">tc</span><span class="p">),</span> <span class="n">p</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">tc</span><span class="p">,</span> <span class="n">p</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">target_counts</span><span class="p">,</span> <span class="n">probs</span><span class="p">)</span>
    <span class="p">])</span>

    <span class="k">return</span> <span class="n">augmented</span>
</code></pre></div>
<h3 id="2-cell-mixtures-simulates-doublets">2. Cell Mixtures (Simulates doublets)</h3>
<div class="highlight"><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_doublet</span><span class="p">(</span><span class="n">cell1</span><span class="p">,</span> <span class="n">cell2</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Simulate doublet by averaging two cells.&quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">cell1</span> <span class="o">+</span> <span class="n">cell2</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span>
</code></pre></div>
<h2 id="benchmark-tasks">Benchmark Tasks</h2>
<h3 id="1-cell-type-annotation-scrna-seq">1. Cell Type Annotation (scRNA-seq)</h3>
<p><strong>Input</strong>: Gene expression matrix <code>(n_cells, n_genes)</code><br />
<strong>Output</strong>: Cell type labels <code>(n_cells,)</code></p>
<div class="highlight"><pre><span></span><code><span class="c1"># Example: PBMC cell type classification</span>
<span class="n">X_train</span> <span class="o">=</span> <span class="n">adata_train</span><span class="o">.</span><span class="n">X</span>  <span class="c1"># (n_train, n_genes)</span>
<span class="n">y_train</span> <span class="o">=</span> <span class="n">adata_train</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;cell_type&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>  <span class="c1"># (n_train,)</span>

<span class="c1"># Labels: B cells, T cells, Monocytes, NK cells, etc.</span>
</code></pre></div>
<h3 id="2-gene-expression-prediction">2. Gene Expression Prediction</h3>
<p><strong>Input</strong>: Partial gene expression or other modalities<br />
<strong>Output</strong>: Full gene expression profiles</p>
<h3 id="3-variant-effect-prediction">3. Variant Effect Prediction</h3>
<p><strong>Input</strong>: DNA sequences with variants<br />
<strong>Output</strong>: Pathogenicity scores or functional effects</p>
<h3 id="4-sequence-classification">4. Sequence Classification</h3>
<p><strong>Input</strong>: DNA sequences<br />
<strong>Output</strong>: Labels (e.g., promoter vs non-promoter, coding vs non-coding)</p>
<h2 id="quality-control-metrics">Quality Control Metrics</h2>
<h3 id="scrna-seq-qc">scRNA-seq QC</h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Check key QC metrics</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mean genes/cell: </span><span class="si">{</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_genes&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mean UMIs/cell: </span><span class="si">{</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;n_counts&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="si">:</span><span class="s2">.0f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Mean % mito: </span><span class="si">{</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;percent_mito&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>

<span class="c1"># Recommended thresholds:</span>
<span class="c1"># - n_genes: 200 - 5000</span>
<span class="c1"># - percent_mito: &lt; 5-10%</span>
<span class="c1"># - n_counts: &gt; 500</span>
</code></pre></div>
<h3 id="variant-data-qc">Variant Data QC</h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Check missing rate</span>
<span class="n">missing_rate</span> <span class="o">=</span> <span class="p">(</span><span class="n">genotypes</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Missing genotype rate: </span><span class="si">{</span><span class="n">missing_rate</span><span class="o">*</span><span class="mi">100</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">%&quot;</span><span class="p">)</span>

<span class="c1"># Filter variants by minor allele frequency (MAF)</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">allel</span><span class="w"> </span><span class="kn">import</span> <span class="n">GenotypeArray</span>

<span class="n">gt</span> <span class="o">=</span> <span class="n">GenotypeArray</span><span class="p">(</span><span class="n">genotypes</span><span class="p">)</span>
<span class="n">allele_counts</span> <span class="o">=</span> <span class="n">gt</span><span class="o">.</span><span class="n">count_alleles</span><span class="p">()</span>
<span class="n">maf</span> <span class="o">=</span> <span class="n">allele_counts</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">/</span> <span class="n">allele_counts</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="c1"># Keep MAF &gt; 0.01</span>
<span class="n">keep_variants</span> <span class="o">=</span> <span class="n">maf</span> <span class="o">&gt;</span> <span class="mf">0.01</span>
</code></pre></div>
<h2 id="data-formats">Data Formats</h2>
<h3 id="anndata-structure">AnnData Structure</h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Recommended AnnData structure for benchmarking</span>
<span class="n">adata</span><span class="o">.</span><span class="n">X</span>  <span class="c1"># Main data matrix (can be sparse)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span>  <span class="c1"># Cell/sample metadata (pandas DataFrame)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">var</span>  <span class="c1"># Gene/feature metadata (pandas DataFrame)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">layers</span>  <span class="c1"># Alternative representations (raw, scaled, etc.)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsm</span>  <span class="c1"># Multi-dimensional annotations (PCA, UMAP)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">varm</span>  <span class="c1"># Multi-dimensional gene annotations</span>
<span class="n">adata</span><span class="o">.</span><span class="n">uns</span>  <span class="c1"># Unstructured metadata (dataset info)</span>
</code></pre></div>
<h3 id="example-dataset-metadata">Example Dataset Metadata</h3>
<div class="highlight"><pre><span></span><code><span class="nt">dataset_id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">pbmc_68k</span>
<span class="nt">name</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">PBMC 68k (10x Genomics)</span>
<span class="nt">modality</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">scRNA-seq</span>
<span class="nt">n_cells</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">68579</span>
<span class="nt">n_genes</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">20387</span>
<span class="nt">technology</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">10x Chromium v2</span>
<span class="nt">organism</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Homo sapiens</span>
<span class="nt">tissue</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Peripheral blood mononuclear cells (PBMC)</span>
<span class="nt">preprocessing</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Scanpy 1.9.1</span>
<span class="nt">normalization</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">log1p(CPM)</span>
<span class="nt">feature_selection</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">top_2000_HVG</span>
<span class="nt">reference</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Zheng et al. (2017)</span>
</code></pre></div>
<h2 id="foundation-model-input-formats">Foundation Model Input Formats</h2>
<h3 id="for-geneformer-and-similar-models">For Geneformer and Similar Models</h3>
<div class="highlight"><pre><span></span><code><span class="c1"># Geneformer expects rank-ordered gene tokens</span>
<span class="k">def</span><span class="w"> </span><span class="nf">prepare_geneformer_input</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">2048</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Convert AnnData to Geneformer input format.</span>

<span class="sd">    Returns ranked gene indices per cell.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Rank genes by expression within each cell</span>
    <span class="n">cell_inputs</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">n_obs</span><span class="p">):</span>
        <span class="n">cell_expr</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">toarray</span><span class="p">()</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>

        <span class="c1"># Rank genes (descending)</span>
        <span class="n">ranked_genes</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">cell_expr</span><span class="p">)[::</span><span class="o">-</span><span class="mi">1</span><span class="p">][:</span><span class="n">n_genes</span><span class="p">]</span>

        <span class="n">cell_inputs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ranked_genes</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">cell_inputs</span><span class="p">)</span>  <span class="c1"># (n_cells, n_genes)</span>
</code></pre></div>
<h3 id="for-dna-sequence-models-dnabert-hyenadna-etc">For DNA Sequence Models (DNABERT, HyenaDNA, etc.)</h3>
<div class="highlight"><pre><span></span><code><span class="c1"># K-mer tokenization for DNABERT</span>
<span class="k">def</span><span class="w"> </span><span class="nf">kmer_tokenize</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">stride</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Tokenize DNA sequence into k-mers.</span>

<span class="sd">    Args:</span>
<span class="sd">        sequence: DNA string</span>
<span class="sd">        k: K-mer size</span>
<span class="sd">        stride: Step size</span>

<span class="sd">    Returns:</span>
<span class="sd">        tokens: List of k-mer strings</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">tokens</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span> <span class="o">-</span> <span class="n">k</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">stride</span><span class="p">):</span>
        <span class="n">tokens</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">sequence</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="n">k</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">tokens</span>

<span class="c1"># Example</span>
<span class="n">seq</span> <span class="o">=</span> <span class="s2">&quot;ATCGATCGATCG&quot;</span>
<span class="n">tokens</span> <span class="o">=</span> <span class="n">kmer_tokenize</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">stride</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tokens</span><span class="p">)</span>  <span class="c1"># [&#39;ATCGAT&#39;, &#39;GATCGA&#39;, &#39;TCGATC&#39;]</span>
</code></pre></div>
<h2 id="itu-ai4h-alignment">ITU AI4H Alignment</h2>
<p>This specification aligns with:</p>
<ul>
<li><strong>DEL3 Section 4.2</strong>: Data format requirements for AI4H benchmarks</li>
<li><strong>DEL0.1</strong>: Genomics terminology standardization</li>
<li><strong>FAIR principles</strong>: Findable, Accessible, Interoperable, Reusable data</li>
</ul>
<h2 id="tools-libraries">Tools &amp; Libraries</h2>
<h3 id="scrna-seq">scRNA-seq</h3>
<ul>
<li><strong>Scanpy</strong>: Single-cell analysis in Python</li>
<li><strong>Seurat</strong>: R toolkit for scRNA-seq</li>
<li><strong>scVI-tools</strong>: Probabilistic models for single-cell omics</li>
</ul>
<h3 id="genomics">Genomics</h3>
<ul>
<li><strong>BioPython</strong>: Sequence analysis</li>
<li><strong>PyVCF / scikit-allel</strong>: Variant data handling</li>
<li><strong>pybedtools</strong>: Genomic interval operations</li>
</ul>
<h3 id="foundation-model-libraries">Foundation Model Libraries</h3>
<ul>
<li><strong>Geneformer</strong>: Transformer for scRNA-seq</li>
<li><strong>DNABERT</strong>: BERT for DNA sequences</li>
<li><strong>Nucleotide Transformer</strong>: Multi-species genomic foundation model</li>
</ul>
<h2 id="references">References</h2>
<ol>
<li>Wolf, F. A., et al. (2018). SCANPY: large-scale single-cell gene expression data analysis. <em>Genome Biology</em>, 19(1), 15.</li>
<li>Theodoris, C. V., et al. (2023). Transfer learning enables predictions in network biology. <em>Nature</em>, 618, 616-624. (Geneformer)</li>
<li>Ji, Y., et al. (2021). DNABERT: pre-trained Bidirectional Encoder Representations from Transformers model for DNA-language in genome. <em>Bioinformatics</em>, 37(15), 2112-2120.</li>
</ol>
<h2 id="related-documentation">Related Documentation</h2>
<ul>
<li><a href="../fmri/">fMRI Specifications</a></li>
<li><a href="../smri/">sMRI Specifications</a></li>
<li><a href="../../analysis_recipes/prediction_baselines/">Prediction Baselines</a></li>
</ul>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      
      <script id="__config" type="application/json">{"annotate": null, "base": "../../..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.sections", "navigation.top", "navigation.tracking", "navigation.instant", "navigation.expand", "toc.integrate", "content.code.copy", "search.suggest", "search.highlight"], "search": "../../../assets/javascripts/workers/search.7a47a382.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.e71a0d61.min.js"></script>
      
    
  </body>
</html>